### DISCUSSION TITLE
ProjectInto unpredictable behavior - only selected properties populated

### DISCUSSION THREAD

**Discussion by mic2zar** [ravendb/ravendb#16528] [Q&A] [answered]
I am in the process of optimizing our database. We were in most of our indexes storing all the fields and querying was working like a charm. Now, we give up on storing and try to use ProjectInto instead [1]. But it has some unpredictable behavior. Some of the properties of indexes are available and some are not and set to null. Documentation is silent on this topic. Could anybody please let me know, why some of the properties are not getting populated (what’s the logic or what shall we enable to get all of them populated). Screenshots of the problem from the code and in studio below:
 
![query_in_code](https://github.com/ravendb/ravendb/assets/6518280/420f33e4-413e-4b6b-a783-3b44e74e1422)

![query_in_studio](https://github.com/ravendb/ravendb/assets/6518280/2b561b8f-5fb0-4712-aed4-dc441da538c2)

[1] [Querying: Projections | RavenDB 5.4 Documentation](https://ravendb.net/docs/article-page/5.4/csharp/indexes/querying/projections#projectinto)

**Comment by grisha-kotler** [comment]
Hi,

The projected fields are taken from the index (if you stored the fields), if not, it's taken from the document.

https://ravendb.net/docs/article-page/5.3/csharp/client-api/session/querying/how-to-project-query-results#projectinto
See the note at the bottom:
If the projected fields are stored inside the index itself (`FieldStorage.Yes` in the index definition), then the query results will be created directly from there instead of retrieving documents in order to project.

**Reply by mic2zar** [reply]
Thank you for your answer. Please let me understand this logic better. Here is my scenario and my current understanding:

1) I do **not** store fields in index at all (FieldStorage.Yes is not set)
2) If the field with the **identical** name **exists** in the original indexed document, it will be populated with ProjectInto (in my example above it was field OwnerId)
3) If the field with the **identical** name **does not exists** in the original document, even if it is just a simple mapping (in my example above Id --> LessonId) it will be not populated with ProjectInto and its value in result will remain null
4) Any other complex case (e.g. selecting and aggregating some values from the collection) will always require to store this field in index otherwise it cannot be retrieved during querying. ProjectInto can only be applied if the field is stored.

**Comment by grisha-kotler** [comment, marked as answer]
You are correct.

If you don't store the fields (FieldStorage.Yes is not set), then you cannot retrieve it.

You can project any field from the document (assuming that those fields exist in the document).
For example:

```
from index 'Lessons' as l
select {
    LeesonId: l.Lesson + " - " + l.MyLesson,
    OwnerId: l.OwnerId,
    Count: l.Count * 3
}
```
